streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java

```diff
@@ -0,0 +1,208 @@
+
+    @Override
+    public boolean assign(final Map<UUID, ClientState> clients,
+                          final Set<TaskId> allTaskIds,
+                          final Set<TaskId> statefulTaskIds,
+                          final AssignorConfiguration.AssignmentConfigs configs) {
+        final int numStandbyReplicas = configs.numStandbyReplicas;
+        final Set<String> rackAwareAssignmentTags = new HashSet<>(configs.rackAwareAssignmentTags);
+        final Map<TaskId, UUID> statefulTasksWithClients = new HashMap<>();
+
+        statefulTaskIds.forEach(statefulTaskId -> clients.forEach((uuid, clientState) -> {
+            if (clientState.activeTasks().contains(statefulTaskId)) {
+                statefulTasksWithClients.put(statefulTaskId, uuid);
+            }
+        }));
+
+        final Map<TaskId, Integer> tasksToRemainingStandbys = computeTasksToRemainingStandbys(numStandbyReplicas,
+                                                                                              statefulTasksWithClients);
+
+        statefulTasksWithClients.forEach((taskId, clientId) -> assignStandbyTasksForActiveTask(numStandbyReplicas,
+                                                                                               taskId,
+                                                                                               clientId,
+                                                                                               rackAwareAssignmentTags,
+                                                                                               clients,
+                                                                                               tasksToRemainingStandbys));
+
+        return true;
+    }
+
+    @Override
+    public boolean isAllowedTaskMovement(final ClientState source, final ClientState destination) {
+        final Map<String, String> sourceClientTags = source.clientTags();
+        final Map<String, String> destinationClientTags = destination.clientTags();
+
+        for (final Entry<String, String> sourceClientTagEntry : sourceClientTags.entrySet()) {
+            if (!sourceClientTagEntry.getValue().equals(destinationClientTags.get(sourceClientTagEntry.getKey()))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static Map<TaskId, Integer> computeTasksToRemainingStandbys(final int numStandbyReplicas, final Map<TaskId, UUID> statefulTasksWithClients) {
+        return statefulTasksWithClients.keySet().stream().collect(toMap(Function.identity(), t -> numStandbyReplicas));
+    }
+
+    private static void fillClientsTagStatistics(final Map<UUID, ClientState> clientStates,
+                                                 final Map<String, Set<UUID>> clientsPerTagValue,
+                                                 final Map<String, Set<String>> tagKeyToTagValuesMapping) {
```

Comment:
IMO, the code is easier readable if you name the variables consistently like `tagValueToClients` and `tagKeyToTagValues` or `clientsForTagValue` and `tagValuesForTagKey`. I prefer the former because it better visualises the mapping, but that is a matter of taste, I guess.

Reply:
Done